File: scala/rational.scala
class Rational(n: Int, d: Int) {
  require(d != 0)

  private val g  = gcd(n.abs, d.abs)
  val numer: Int = n / g
  val denom: Int = d / g

  def this(n: Int) = this(n, 1) // auxiliary constructor

  override def toString = numer + "/" + denom

  def + (that: Rational): Rational =
    new Rational(
      numer * that.denom + that.numer * denom,
      denom * that.denom
    )

  def * (that: Rational): Rational =
    new Rational(
      numer * that.numer,
      denom * that.denom
    )

  def lessThan(that: Rational) =
    this.numer * that.denom < that.numer * this.denom

  def max(that: Rational) =
    if (this.lessThan(that)) that else this

  // private methods
  private def gcd(a: Int, b: Int): Int =
    if (b == 0) a else gcd(b, a % b)
}

File: ruby/rational.rb
class Rational
  attr_reader :numer, :denom

  def initialize(numer, denom=1)
    @numer, @denom = numer, denom
  end

  def to_string
    "<#Rational #{self.object_id}: #{numer} / #{denom} >"
  end

  def add(that)
    Rational.new(
      self.numer * that.denom + that.numer * self.denom,
      self.denom * that.denom )
  end
  alias :+, :add

  def multiply(that)
    Rational.new(
      self.numer * that.numer,
      this.denom * that.denom )
  end
  alias :*, :multiply

  def lessThan?(that)
    self.numer * that.denom < that.numer * self.denom
  end
  alias :<, :lessThan?

  def max(that)
    self.lessThan?(that) ? that : self
  end

  private

  def gcd(thisInt, thatInt)
    thatInt == 0 ? a : gcd(thatInt, thisInt % thatInt) 
  end
end
